client/landing/stepper/declarative-flow/internals/components/step-route/index.tsx

```diff
@@ -11,9 +28,92 @@ type StepRouteProps = {
 	renderStep: ( step: StepperStep ) => JSX.Element | null;
 };
 
+//TODO: Check we can move RenderStep function to here and remove the renderStep prop
 const StepRoute = ( { step, flow, showWooLogo, renderStep }: StepRouteProps ) => {
+	const userIsLoggedIn = useSelector( isUserLoggedIn );
+	const loginUrl = useLoginUrlForFlow( { flow } );
 	const stepContent = renderStep( step );
+	const { site, siteSlugOrId } = useSiteData();
 
+	// Ensure that the selected site is fetched, if available. This is used for event tracking purposes.
+	// See https://github.com/Automattic/wp-calypso/pull/82981.
+
+	// TODO: Check if it is duplicated with the useSiteData hook
+	const selectedSite = useSelector( ( state ) => site && getSite( state, siteSlugOrId ) );
+	const isRequestingSelectedSite = useSelector(
+		( state ) => site && isRequestingSite( state, siteSlugOrId )
+	);
+	const intent = useIntent();
+	const design = useSelectedDesign();
+
+	// Short-circuit this if the site slug or ID is not available.
+	const hasRequestedSelectedSite = siteSlugOrId
+		? !! selectedSite && ! isRequestingSelectedSite
+		: true;
+
+	useEffect( () => {
+		if ( ! step.requiresLoggedInUser || userIsLoggedIn ) {
+			return;
+		}
+
+		window.location.assign( loginUrl );
+	}, [ step, loginUrl, userIsLoggedIn ] );
+
+	//TODO: Move this logic to a separate hook
+	useEffect( () => {
+		// We record the event only when the step is not empty. Additionally, we should not fire this event whenever the intent is changed
+		if ( ! hasRequestedSelectedSite ) {
+			return;
+		}
+
+		// If we're redirecting in the effect above, skip any step or page tracking.
+		if ( step.requiresLoggedInUser && ! userIsLoggedIn ) {
+			return;
+		}
+
+		if ( ! stepContent ) {
+			return;
+		}
+
+		const signupCompleteFlowName = getSignupCompleteFlowNameAndClear();
+		const signupCompleteStepName = getSignupCompleteStepNameAndClear();
+
+		const isReEnteringStep =
+			signupCompleteFlowName === flow.name && signupCompleteStepName === step.slug;
+
+		if ( ! isReEnteringStep ) {
+			recordStepStart( flow.name, kebabCase( step.slug ), {
+				intent,
+				is_in_hosting_flow: isAnyHostingFlow( flow.name ),
+				...( design && { assembler_source: getAssemblerSource( design ) } ),
+			} );
+
+			const stepOldSlug = getStepOldSlug( step.slug );
+
+			if ( stepOldSlug ) {
+				recordStepStart( flow.name, kebabCase( stepOldSlug ), {
+					intent,
+					is_in_hosting_flow: isAnyHostingFlow( flow.name ),
+					...( design && { assembler_source: getAssemblerSource( design ) } ),
+				} );
+			}
+		}
+
+		// Also record page view for data and analytics
+		const pathname = window.location.pathname || '';
```

Comment:
Can `pathname` ever be falsy?

Reply:
You're right -- this is probably unnecessary -- I think `location.pathname` will always be a string.

Reply:
Removed!